File: src/Paper.cpp
#include "const.h"
#include <iomanip>

Paper::Paper() {

    width  = height = f_v = M = 0;
    grains = NULL;
    g = NULL;

}

void Paper::init(int w, int h, float fv, float pw, float pm) {
    
    width  = w;
    height = h;
    f_v    = fv;
    W      = pw;
    M      = pm;
    grains = new Grain* [w];
    for (int i = 0; i < w; ++i)
        grains[i] = new Grain[h];

    initAllGrains();

}

void Paper::initAllGrains() {

    srand(time(NULL));
    for (int x = 1; x < width - 1; ++x)
        for (int y = 1; y < height - 1; ++y) {
            grains[x][y].updateH(rand() % 1001 / 1000.0);
            grains[x][y].init(&grains[x-1][y], &grains[x][y-1], &grains[x-1][y-1],
                              &grains[x+1][y], &grains[x][y+1], &grains[x+1][y+1]);
        }

}

void Paper::setPosition(int x, int y) {

    int posX = x, posY = y;
    if (x < 2)
        posX = 2;
    else if (x > width - 2) 
        posX = width - 3;
    if (y < 2) 
        posY = 2;
    else if (y > height - 2) 
        posY = height - 3;
    g = &grains[posX][posY];

}

void Paper::show(float g_p) {
//void Paper::show() {
 /* 
    for (int x = 2; x < width - 2; ++x)
        for (int y = 2; y < height - 2; ++y) {
            cout << "(" << x << "," << y << ")->";
            setPosition(x, y);
            cout << setw(9) << g->getH() << ","
                 << setw(9) << g->getH_max() << ","
                 << setw(9) << g->getH_min() << ","
                 << setw(9) << g->getH_sum() << "#"
                 << setw(9) << g->getHp_max() << ","
                 << setw(9) << g->getHp_min() << ","
                 << setw(9) << g->getHp_sum() << "#"
                 << setw(9) << g->getD_k(g->getH()) << ","
                 << setw(9) << g->getT_v(f_v) << ","
                 << setw(9) << g->getB_k(M, p, f_v) << ","
                 << setw(9) << g->getT_k() << endl;
        }      
*/
  
    for(int i=0; i < width; ++i) 
        for (int j = 0; j < height; ++j){
            float grey = 1 - grains[i][j].getT_k() * g_p / 1000;
            glColor3f(grey, grey, grey);
            glBegin(GL_POINTS);
                glVertex2i(i, j);
            glEnd();
        }

}

File: web/src/paper.js
/*
 * Paper
 */

function Paper(w, h, fv, pw, pm) {

  this.width  = w;
  this.height = h;
  this.f_v    = fv;
  this.W      = pw;
  this.M      = pm;

  this.grains = (function(w, h) {

    var gs = new Array(w);

    for (var i = 0; i < w; ++i) {
      gs[i] = new Array(h);
      for (var j = 0; j < h; ++j) 
        gs[i][j] = new Grain(Math.random());    
    }

    return gs;
  
  })(w, h);

}

Paper.prototype = {
  
  getW: function() { return this.W; },
  getM: function() { return this.M; },

  updateH: function(i, j, gh) {
    this.grains[i][j].updateH(gh);
  },

  updateT: function(i, j, gt) {
    this.grains[i][j].updateT(gt);
  },

  getH: function(i, j) {
    return this.grains[i][j].h;
  },

  getH_max: function(i, j) {

    return Math.max(
      this.grains[i][j].h,
      this.grains[i+1][j].h,
      this.grains[i+1][j+1].h,
      this.grains[i][j+1].h
    );

  },

  getH_min: function(i, j) {

    return Math.min(
      this.grains[i][j].h,
      this.grains[i+1][j].h,
      this.grains[i+1][j+1].h,
      this.grains[i][j+1].h
    );

  },

  getH_sum: function(i, j) {

    return Utils.sum(
      this.grains[i][j].h,
      this.grains[i+1][j].h,
      this.grains[i+1][j+1].h,
      this.grains[i][j+1].h
    );

  },

  getHp_max: function(i, j) {
    return this.getH_max(i, j) - this.getH_min(i, j);
  },

  getHp_min: function(i, j) {
    return 0;
  },

  getHp_sum: function(i, j) {

    var h_max = this.getH_max(i, j);
    return h_max - this.grains[i][j].h     +
           h_max - this.grains[i+1][j].h   + 
           h_max - this.grains[i+1][j+1].h + 
           h_max - this.grains[i][j+1].h;

  },

  getD_k: function(i, j, gh) {

    h_sum = this.getH_sum(i, j);
    if (h_sum < Const.ZERO)
      return 0;
    return gh / h_sum;

  },

  getT_v: function(i, j) {
    return this.f_v * this.getHp_sum(i, j) / 4.0;
  },

  getT_k: function(i, j) {
    return this.grains[i][j].getT_k();
  },

  getL_k: function(i, j) {

    var h = this.grains[i][j].h;
    return this.getD_k(i-1, j, h) * this.getT_v(i-1, j) +
           this.getD_k(i, j-1, h) * this.getT_v(i, j-1) +
           this.getD_k(i-1, j-1, h) * this.getT_v(i-1, j-1) +
           this.getD_k(i, j, h) * this.getT_v(i, j);

  },

  getB_k: function(i, j, p) {

    var h = this.grains[i][j].h;
    var b_k = this.getD_k(i-1, j, h) * this.grains[i-1][j].b +
              this.getD_k(i, j-1, h) * this.grains[i][j-1].b +
              this.getD_k(i-1, j-1, h) * this.grains[i-1][j-1].b +
              this.getD_k(i, j, h) * this.grains[i][j].b;

    return this.grains[i][j].getT_k() > this.getL_k(i, j) ? b_k * Math.pow(this.M, p) : b_k;

  },

  setB_v: function(i, j, bv) {
    this.grains[i][j].setB_v(bv);
  }

}
